Flaky `TestLambdaInDebugInfo`

Introduced by #4207, ref. https://github.com/nspcc-dev/neo-go/pull/4207#issuecomment-4767424960.
```
2026-06-19T08:53:08.1596052Z === RUN   TestLambdaInDebugInfo
2026-06-19T08:53:08.1596392Z     lambda_test.go:107: 
2026-06-19T08:53:08.1597456Z         	Error Trace:	/home/runner/work/neo-go/neo-go/pkg/compiler/lambda_test.go:107
2026-06-19T08:53:08.1597844Z         	Error:      	Received unexpected error:
2026-06-19T08:53:08.1598414Z         	            	invalid label target: -1 at 16
2026-06-19T08:53:08.1598739Z         	Test:       	TestLambdaInDebugInfo
2026-06-19T08:53:08.1598935Z --- FAIL: TestLambdaInDebugInfo (0.19s)
```

--- a/pkg/compiler/codegen.go
+++ b/pkg/compiler/codegen.go
@@ -519,8 +519,8 @@ func (c *codegen) convertDeployFuncs() int {
 
 func (c *codegen) convertFuncDecl(file ast.Node, decl *ast.FuncDecl, pkg *types.Package) *funcScope {
 	var (
-		f            *funcScope
-		ok, isLambda bool
+		f  *funcScope
+		ok bool
 	)
 	isInit := isInitFunc(decl)
 	isDeploy := isDeployFunc(decl)
@@ -543,7 +543,6 @@ func (c *codegen) convertFuncDecl(file ast.Node, decl *ast.FuncDecl, pkg *types.
 				panic("ICE: lambda must already be marked as compiled")
 			}
 			f = l.funcScope
-			isLambda = ok
 			c.setLabel(f.label)
 		} else {
 			f = c.newFunc(decl)
@@ -626,17 +625,15 @@ func (c *codegen) convertFuncDecl(file ast.Node, decl *ast.FuncDecl, pkg *types.
 
 	f.rng.End = uint16(c.prog.Len() - 1)
 
-	if !isLambda {
-		for _, f := range c.lambda {
-			if f.compiled {
-				continue
-			}
-			if _, ok := c.lambda[c.getIdentName("", f.decl.Name.Name)]; !ok {
-				panic("ICE: lambda name doesn't match map key")
-			}
-			f.compiled = true
-			c.convertFuncDecl(file, f.decl, pkg)
+	for _, f := range c.lambda {
+		if f.compiled {
+			continue
 		}
+		if _, ok := c.lambda[c.getIdentName("", f.decl.Name.Name)]; !ok {
+			panic("ICE: lambda name doesn't match map key")
+		}
+		f.compiled = true
+		c.convertFuncDecl(file, f.decl, pkg)
 	}
 
 	if !isInit && !isDeploy {

--- a/pkg/compiler/lambda_test.go
+++ b/pkg/compiler/lambda_test.go
@@ -160,6 +160,86 @@ func TestLambdaInDebugInfo(t *testing.T) {
 				},
 			},
 		},
+		{
+			name: "three-level nested lambdas",
+			src: `package main
+				import "github.com/nspcc-dev/neo-go/pkg/interop/runtime"
+				func PublicContractMethod() {
+					f := func() {
+						g := func() {
+							h := func() {
+								runtime.Log("deep")
+							}
+							h()
+						}
+						g()
+					}
+					f()
+				}
+			`,
+			seqPoints: map[string][]compiler.DebugSeqPoint{
+				"PublicContractMethod": {
+					{StartLine: 4, StartCol: 11, EndLine: 4, EndCol: 17},
+					{StartLine: 4, StartCol: 6, EndLine: 4, EndCol: 11},
+					{StartLine: 13, StartCol: 6, EndLine: 13, EndCol: 9},
+				},
+				"lambda@56": {
+					{StartLine: 5, StartCol: 12, EndLine: 5, EndCol: 18},
+					{StartLine: 5, StartCol: 7, EndLine: 5, EndCol: 12},
+					{StartLine: 11, StartCol: 7, EndLine: 11, EndCol: 10},
+				},
+				"lambda@57": {
+					{StartLine: 6, StartCol: 13, EndLine: 6, EndCol: 19},
+					{StartLine: 6, StartCol: 8, EndLine: 6, EndCol: 13},
+					{StartLine: 9, StartCol: 8, EndLine: 9, EndCol: 11},
+				},
+				"lambda@58": {
+					{StartLine: 7, StartCol: 9, EndLine: 7, EndCol: 28},
+					{StartLine: 60, StartCol: 2, EndLine: 60, EndCol: 63},
+				},
+			},
+		},
+		{
+			name: "lambda returns lambda returns lambda",
+			src: `package main
+				import "github.com/nspcc-dev/neo-go/pkg/interop/runtime"
+				func PublicContractMethod() {
+					f := func() func() func() {
+						return func() func() {
+							return func() {
+								runtime.Log("deep")
+							}
+						}
+					}
+					g := f()
+					h := g()
+					h()
+				}
+			`,
+			seqPoints: map[string][]compiler.DebugSeqPoint{
+				"PublicContractMethod": {
+					{StartLine: 4, StartCol: 11, EndLine: 4, EndCol: 31},
+					{StartLine: 4, StartCol: 6, EndLine: 4, EndCol: 11},
+					{StartLine: 11, StartCol: 11, EndLine: 11, EndCol: 14},
+					{StartLine: 11, StartCol: 6, EndLine: 11, EndCol: 11},
+					{StartLine: 12, StartCol: 11, EndLine: 12, EndCol: 14},
+					{StartLine: 12, StartCol: 6, EndLine: 12, EndCol: 11},
+					{StartLine: 13, StartCol: 6, EndLine: 13, EndCol: 9},
+				},
+				"lambda@56": {
+					{StartLine: 5, StartCol: 14, EndLine: 5, EndCol: 27},
+					{StartLine: 5, StartCol: 7, EndLine: 5, EndCol: 13},
+				},
+				"lambda@58": {
+					{StartLine: 6, StartCol: 15, EndLine: 6, EndCol: 21},
+					{StartLine: 6, StartCol: 8, EndLine: 6, EndCol: 14},
+				},
+				"lambda@60": {
+					{StartLine: 7, StartCol: 9, EndLine: 7, EndCol: 28},
+					{StartLine: 60, StartCol: 2, EndLine: 60, EndCol: 63},
+				},
+			},
+		},
 	}
 	for _, tc := range testCases {
 		t.Run(tc.name, func(t *testing.T) {

